tests/by-util/test_groups.rs

```diff
@@ -1,56 +1,174 @@
+        println!("{}\ntest skipped", version_check_string);
         return;
-    };
+    }
+    let test_users = ["root", "man", "postfix", "sshd", &whoami()];
+
+    let result = new_ucmd!().args(&test_users).run();
+    let exp_result = unwrap_or_return!(expected_result(&test_users));
 
-    // TODO: stdout should be in the form: "username : group1 group2 group3"
+    result
+        .stdout_is(exp_result.stdout_str())
+        .stderr_is(exp_result.stderr_str())
+        .code_is(exp_result.code());
+}
 
-    scene
-        .ucmd()
-        .arg(&username)
-        .succeeds()
-        .stdout_is(expected_result(&[&username]));
+fn check_coreutil_version(util_name: &str, version_expected: &str) -> String {
+    // example:
+    // $ id --version | head -n 1
+    // id (GNU coreutils) 8.32.162-4eda
+    let scene = TestScenario::new(util_name);
+    let version_check = scene
+        .cmd_keepenv(&util_name)
+        .env("LC_ALL", "C")
+        .arg("--version")
+        .run();
+    version_check
+        .stdout_str()
+        .split('\n')
+        .collect::<Vec<_>>()
+        .get(0)
+        .map_or_else(
+            || format!("{}: unexpected output format for reference coreutil: '{} --version'", UUTILS_WARNING, util_name),
+            |s| {
+                if s.contains(&format!("(GNU coreutils) {}", version_expected)) {
+                    s.to_string()
+                } else if s.contains("(GNU coreutils)") {
+                    let version_found = s.split_whitespace().last().unwrap()[..4].parse::<f32>().unwrap_or_default();
+                    let version_expected = version_expected.parse::<f32>().unwrap_or_default();
+                    if version_found > version_expected {
+                    format!("{}: version for the reference coreutil '{}' is higher than expected; expected: {}, found: {}", UUTILS_INFO, util_name, version_expected, version_found)
+                    } else {
+                    format!("{}: version for the reference coreutil '{}' does not match; expected: {}, found: {}", UUTILS_WARNING, util_name, version_expected, version_found) }
+                } else {
+                    format!("{}: no coreutils version string found for reference coreutils '{} --version'", UUTILS_WARNING, util_name)
+                }
+            },
+        )
 }
 
+#[allow(clippy::needless_borrow)]
 #[cfg(unix)]
-fn expected_result(args: &[&str]) -> String {
-    // We want to use GNU id. On most linux systems, this is "id", but on
-    // bsd-like systems (e.g. FreeBSD, MacOS), it is commonly "gid".
-    #[cfg(any(target_os = "linux"))]
-    let util_name = "id";
-    #[cfg(not(target_os = "linux"))]
-    let util_name = "gid";
-
-    TestScenario::new(util_name)
+fn expected_result(args: &[&str]) -> Result<CmdResult, String> {
+    #[cfg(target_os = "linux")]
+    let util_name = util_name!();
+    #[cfg(all(unix, not(target_os = "linux")))]
+    let util_name = &format!("g{}", util_name!());
```

Comment:
The reason this is needed is that in some environments, the GNU coreutils are prefixed with "g".
Originally I wrote this for macOS where the coreutils installed with:
https://github.com/uutils/coreutils/blob/d3181fa60e63569513b5b8cde3a51376b2916d51/.github/workflows/CICD.yml#L272
are prefixed with "g" to not interfere with the BSD counterparts already on the system.

I changed this to "not linux" when I was told that this prefixing is also done on freebsd
https://github.com/uutils/coreutils/pull/2373#discussion_r647787251

Reply:
I understand the need for the variation, but the construction is opaque, using `format!()` and `util_name!()` rather than simple specification of the names.

```suggestion
    #[cfg(target_os = "linux")]
    let util_name = "groups";
    #[cfg(all(unix, not(target_os = "linux")))]
    let util_name = "ggroups";
```

Reply:
Sorry, I misunderstood what you were referring to with "_opaque_".

I'm using the `expected_result` function with the conditional statement for the prefix not just here.
I'm intentionally not using a hard coded `util_name` because it's easier like this to copy&paste these lines to other uutils.

Ultimately, I plan to reduce duplicate code, to do a refactor of the `expected_result` function and to move it to e.g. `tests/common/`.
Right now there are c&p versions of `expected_result` in test_id, test_pinky, test_stat, test_who, test_group.

Reply:
That sounds like a reasonable plan. Just leave a ToDO comment as a prompt for the future refactor; something like...
```suggestion
    // ToDO: [2021-06; jhscheer] refactor this as `let util_name = host_name_for(util_name!())` when that function is added to 'tests/common'
    #[cfg(target_os = "linux")]
    let util_name = util_name!();
    #[cfg(all(unix, not(target_os = "linux")))]
    let util_name = &format!("g{}", util_name!());
```

Reply:
That's a very good idea!